### PR TITLE
Assigned DG Edits

### PR DESCRIPTION
**Pull request by HM33-Stu**
Compilation of HM33-Stu (Hari) Assigned DG Edits

1. Manual Testing For Add and Delete Internships
2. Appendix: Design Decisions
3. User Cases For All Features
4. Implementation For Add/Delete (Optional Feature Discussion Included in Add)

Use case (Hari)
Use case:
Add an internship
Edit an internship
Delete an internship
Sort internship
Find internships
Add task
Delete task
Set deadline
Open detailed view

MSS & Extensions

### CODE REVIEW

File: docs/DeveloperGuide.md [hunk 1]
```diff
@@ -161,6 +163,67 @@ Classes used by multiple components are in the `seedu.addressbook.commons` packa
...
+The transformation of certain fields in the Internship class to optional types is a crucial step in enhancing the flexibility and robustness of the internship application.
+This report outlines the technical details of the implementation, including the different states, usage scenarios, design considerations, alternatives, and their implications.
+
+<img src="images/OptionalFeatures.png" width="280" />
+
+#### Implementation Details (How Is It Implemented):
+
+**Field Conversion:**
+Fields such as location and role have been converted to optional types (where default values have been set should the user not input any values for them)
+This transformation allows these fields to be nullable, providing flexibility in representing internship data.
```
Comment (jieqiboh) [inline docs/DeveloperGuide.md:177]: can remove this line since not nullable

File: docs/DeveloperGuide.md [hunk 2]
```diff
@@ -301,169 +364,208 @@ Priorities: High (Must-Have) - `* * *`, Medium (Nice-To-Have) - `* *`, Low (Not
...
 **MSS**
-
-1. User accesses the main page. 
-2. User selects the option to enter internship information. 
-3. System prompts the user to input internship details such as company name, role title, description, etc. 
-4. User inputs the required internship details. 
-5. System validates the input data. 
-6. System saves the internship information. 
+1. User accesses the main page.
+2. User selects the option to add a new internship.
```
Comment (LWS49) [inline docs/DeveloperGuide.md:371]: select option is not an accurate phrasing
maybe "User inputs the add command with internship details such as company name, contact name, etc."

File: docs/DeveloperGuide.md [hunk 3]
```diff
@@ -161,6 +163,67 @@ Classes used by multiple components are in the `seedu.addressbook.commons` packa
...
+
+**Equality Comparison:**
+The equals() and isSameInternship() methods have been updated to compare optional fields properly, ensuring consistency in object comparison.
+
+**String Representation:**
+The toString() method has been updated to handle optional fields gracefully, providing a comprehensive view of internship details.
+
+**Different States:**
+
+Initial State: The internship application starts with an initial state where all fields are optional and nullable.
```
Comment (HM33-Stu) [inline docs/DeveloperGuide.md:190]: Note!

Comment (jieqiboh) [inline docs/DeveloperGuide.md:190]: change line since not all fields are optional

File: docs/DeveloperGuide.md [hunk 4]
```diff
@@ -161,6 +163,67 @@ Classes used by multiple components are in the `seedu.addressbook.commons` packa
...
+<img src="images/OptionalFeatures.png" width="280" />
+
+#### Implementation Details (How Is It Implemented):
+
+**Field Conversion:**
+Fields such as location and role have been converted to optional types (where default values have been set should the user not input any values for them)
+This transformation allows these fields to be nullable, providing flexibility in representing internship data.
+
+**Constructor Modification:**
+The constructor of the Internship class now accepts optional parameters for the transformed fields, ensuring compatibility with the new optional field structure.
```
Comment (jieqiboh) [inline docs/DeveloperGuide.md:180]: technically the constructor does not accept optional parameters
The parser accepts optional fields and sets default values in the constructor

File: docs/DeveloperGuide.md [hunk 5]
```diff
@@ -161,6 +163,67 @@ Classes used by multiple components are in the `seedu.addressbook.commons` packa
...
+The decision to use optional fields introduces increased flexibility in representing internship data.
+However, this flexibility may come at the cost of slightly reduced performance due to optional value handling.
+
+**Alternative and Its Implications:**
+
+#### Alternative Approach:
+The alternative approach was to directly transform role and location into optional fields using `Optional<>`. <br>
+This was attempted and removed as it did create some issues where `NULL` values were being passed around and created confusions when testing other features.
+
+**Implications:**
```
Comment (jieqiboh) [inline docs/DeveloperGuide.md:213]: This is for undo/redo

File: docs/DeveloperGuide.md [hunk 6]
```diff
@@ -161,6 +163,67 @@ Classes used by multiple components are in the `seedu.addressbook.commons` packa
...
+Pros: Reduced memory usage as only relevant data for undo/redo operations is stored.
+Cons: Increased complexity and potential code duplication as undo/redo logic needs to be implemented for each command.
+This approach may also introduce maintenance challenges and potential errors if not implemented correctly.
```
Comment (HM33-Stu) [inline docs/DeveloperGuide.md:216]: Note to remove

File: docs/DeveloperGuide.md (hunk 2 above)
Comment (LWS49) [reply docs/DeveloperGuide.md:371]: remove steps 2 and 3

File: docs/DeveloperGuide.md [hunk 7]
```diff
@@ -301,169 +364,208 @@ Priorities: High (Must-Have) - `* * *`, Medium (Nice-To-Have) - `* *`, Low (Not
...
+6a. System fails to save the internship information. <br>
+6a1. System displays an error message and prompts the user to retry or cancel the submission.
+
+Use case ends.
+
+**Use Case: Edit Internship Entry From Main Window**
+
+**MSS**
+1. User accesses the main page.
+2. User selects the option to view all internship details.
```
Comment (LWS49) [inline docs/DeveloperGuide.md:394]: remove step 2 and 5
Combine steps 4 and 6 together
"User inputs the internship to be edited and the modifications"

File: docs/DeveloperGuide.md [hunk 8]
```diff
@@ -447,169 +510,172 @@ Priorities: High (Must-Have) - `* * *`, Medium (Nice-To-Have) - `* *`, Low (Not
...
 
-    Use case ends.
+Use case ends.
 
-**Use case: View All Internship Details**
+**Use Case: Add Internship Task From Main Window**
 
 **MSS**
+1. User accesses the main window.
+2. User uses the command to specify which internship entry to add a task to and the task details.
```
Comment (LWS49) [inline docs/DeveloperGuide.md:615]: System validates that task is valid

File: docs/DeveloperGuide.md [hunk 9]
```diff
@@ -447,169 +510,172 @@ Priorities: High (Must-Have) - `* * *`, Medium (Nice-To-Have) - `* *`, Low (Not
...
-3. System retrieves and displays a list of all entered internship details.
-4. User selects the internship entry to which notes will be added.
-5. System presents the selected internship details along with the current notes section.
-6. User enters additional information in the notes section.
-7. System saves the updated notes.
-8. System displays a confirmation message indicating successful addition of notes.
+1a. User is unable to access the main window. <br>
+1a1. System displays an error message. <br>
+2a. System recognizes that user has entered an invalid task index. <br>
+2a1. System displays an error message indicating the index error. <br>
```
Comment (LWS49) [inline docs/DeveloperGuide.md:643]: System recognizes that user has entered an invalid internship index.

File: docs/DeveloperGuide.md [hunk 10]
```diff
@@ -662,23 +728,45 @@ testers are expected to do more *exploratory* testing.
...
+    1. Test case: `add /com Facebook /desc create new recommendation engine /status ongoing /poc sally tan /email hr@facebook.com /phone 90890375`<br>
+       Expected: Internship is added to the list in the Main Window. Details of the internship are shown in the Message Box, Main Window and Detailed Internship View.
+
+    1. Test case: `add /com Google /desc create new recommendation forum /status accepted /poc jane tan /email hr@google.com`<br>
+       Expected: Internship is not added to the list in the Main Window. Details of the internship are not shown in the Message Box, Main Window and Detailed Internship View. <br>
+       Reason: The internship entry is missing some compulsory fields (in this case, the phone number of the contact)
 
-1. Deleting a person while all persons are being shown
+    1. Test case: `add /co Stripe /desc create new application /status to_apply /poc janet sim /email hr@stripe.com`<br>
+       Expected: Internship is not added to the list in the Main Window. Details of the internship are not shown in the Message Box, Main Window and Detailed Internship View. <br>
```
Comment (LWS49) [inline docs/DeveloperGuide.md:748]: Need to say error message show in message box. same for the rest.

File: docs/DeveloperGuide.md [hunk 11]
```diff
@@ -662,23 +728,45 @@ testers are expected to do more *exploratory* testing.
...
+       Reason: The wrong command prefix for company is used (`/co` instead of `/com`)
 
-    1. Prerequisites: List all persons using the `list` command. Multiple persons in the list.
+
+### Deleting an internship
+
+1. Deleting an internship while all internships are being shown or specific internships based on find command
+
+    1. Prerequisites: List all internships using the `list` command. Multiple persons in the list.
```
Comment (LWS49) [inline docs/DeveloperGuide.md:756]: multiple internships in the list.

File: docs/DeveloperGuide.md [hunk 12]
```diff
@@ -662,23 +728,45 @@ testers are expected to do more *exploratory* testing.
...
 
     1. Test case: `delete 1`<br>
-       Expected: First contact is deleted from the list. Details of the deleted contact shown in the status message. Timestamp in the status bar is updated.
+       Expected: First internship is deleted from the list. Details of the deleted internship shown in the message box. Details of the internship are no longer visible in the Main Window.
 
     1. Test case: `delete 0`<br>
-       Expected: No person is deleted. Error details shown in the status message. Status bar remains the same.
+       Expected: No internship is deleted. Error details shown in the message box. Internship remains visible in the Main Window and Detailed Internship View.
```
Comment (LWS49) [inline docs/DeveloperGuide.md:762]: All internships remain visible, since no internship is specified

File: docs/DeveloperGuide.md [hunk 13]
```diff
@@ -662,23 +728,45 @@ testers are expected to do more *exploratory* testing.
...
+
+    1. Test case: `delete -1`<br>
+      Expected: No internship is deleted. Error details shown in the message box. Internship remains visible in the Main Window and Detailed Internship View.
```
Comment (LWS49) [inline docs/DeveloperGuide.md:765]: All internships remain visible, since no internship is specified

File: docs/DeveloperGuide.md [hunk 14]
```diff
@@ -688,6 +776,36 @@ Make sure to use the `exit` command or the close button to save data while closi
...
+--------------------------------------------------------------------------------------------------------------------
+
+## **Appendix: Design Decisions**
+
+### Prefix-based Command Format
+
+In the design of our command-line interface, we made the decision to use a prefix-based format for our commands, such as `/com` for company name, instead of the more conventional `com/`. <br> This decision was made based on several considerations:
+
+1. **Uniqueness**: The use of a leading slash makes our commands distinct and immediately recognizable. This reduces the likelihood of conflicts with other command-line applications or conventions.
+2. **Ease of Parsing**: Having a consistent prefix-based format simplifies the parsing process. The parser can easily identify the start of a parameter by looking for the `/` character.
```
Comment (LWS49) [inline docs/DeveloperGuide.md:788]: ease of parsing i think should remove, lowkey i think ours is harder to parse

File: docs/DeveloperGuide.md [hunk 15]
```diff
@@ -688,6 +776,36 @@ Make sure to use the `exit` command or the close button to save data while closi
...
+<br> The choice of these fields was based on the following considerations:
+
+1. **CompanyName**: The name of the company offering the internship is a crucial identifier. Two internships at different companies are definitely not the same.
+2. **ContactName**: The contact person for the internship could be important if the user needs to communicate with the company. Two internships with different contact persons might imply different points of contact, hence they are not the same.
+3. **ContactEmail**: Similar to `contactName`, the contact email could be a significant identifier as it might imply different points of contact.
+4. **ContactNumber**: The contact number, like the contact name and email, could be a significant identifier for the same reasons.
+5. **ApplicationStatus**: The status of the application (e.g., applied, accepted, rejected) is a crucial piece of information for the user. Two internships with different application statuses are not the same from the user's perspective.
+6. **Description**: The description of the internship could contain important details about the internship. Two internships with different descriptions are not the same.
+
```
Comment (LWS49) [inline docs/DeveloperGuide.md:806]: if possible: explain why location, remark and tasklist are not considered (the latter two explain briefly)

File: docs/DeveloperGuide.md [hunk 16]
```diff
@@ -688,6 +776,36 @@ Make sure to use the `exit` command or the close button to save data while closi
...
+<br> These fields are `companyName`, `contactName`, `contactEmail`, `contactNumber`, `applicationStatus`, and `description`. 
+<br> The choice of these fields was based on the following considerations:
+
+1. **CompanyName**: The name of the company offering the internship is a crucial identifier. Two internships at different companies are definitely not the same.
+2. **ContactName**: The contact person for the internship could be important if the user needs to communicate with the company. Two internships with different contact persons might imply different points of contact, hence they are not the same.
+3. **ContactEmail**: Similar to `contactName`, the contact email could be a significant identifier as it might imply different points of contact.
+4. **ContactNumber**: The contact number, like the contact name and email, could be a significant identifier for the same reasons.
+5. **ApplicationStatus**: The status of the application (e.g., applied, accepted, rejected) is a crucial piece of information for the user. Two internships with different application statuses are not the same from the user's perspective.
```
Comment (LWS49) [inline docs/DeveloperGuide.md:804]: remove applicationstatus

File: docs/DeveloperGuide.md [hunk 17]
```diff
@@ -161,10 +163,56 @@ Classes used by multiple components are in the `seedu.addressbook.commons` packa
...
+Fields such as location and role have been converted to optional types (where default values have been set should the user not input any values for them).
+This provides flexibility in representing internship data.
+
+**Constructor Modification:** <br>
+The constructor for the Internship class consists of 2 optional fields (`role` and `location`), ensuring compatibility with the new optional field structure.
+The add command parser sets default values for the 2 aforementioned optional fields, should the user not input any values.
+
+**Equality Comparison:** <br>
+The equals() and isSameInternship() methods account for optional fields differently based on the scenario:
+1. If two internships have identical /company, /poc, /desc fields and have the same /role and /loc, they will be considered as identical internships.
```
Comment (jieqiboh) [inline docs/DeveloperGuide.md:186]: Is this contact email and contact phone also needed to check for equality?

Comment (HM33-Stu) [reply docs/DeveloperGuide.md:186]: oh mb. i shd have used poc fields instead of /poc

**Comment by HM33-Stu** [comment]
Amendments made:

image
image

File: docs/DeveloperGuide.md [hunk 18]
```diff
@@ -17,13 +17,15 @@ title: Developer Guide
     - [[Proposed] Data archiving](#proposed-data-archiving)
     - [Find](#find-feature)
     - [Sort](#sort-feature)
+   -  [Optional Fields](#optional-fields)
```
Comment (jieqiboh) [inline docs/DeveloperGuide.md:20]: the link doesn't work

Comment (HM33-Stu) [reply docs/DeveloperGuide.md]: Done. Resolved in next commit

**Comment by HM33-Stu** [comment]
image

File: docs/DeveloperGuide.md [hunk 19]
```diff
@@ -526,169 +559,197 @@ Your all-in-one solution for seamless application management.
 
 (For all use cases below, the **System** is `CareerSync` and the **Actor** is the `user`, unless specified otherwise)
 
-**Use case: Enter Internship Information From The Main Page**
+**Use Case: Enter Internship Entry From Main Window**
 
 **MSS**
+1. User accesses the main window.
+2. User uses the command and inputs internship details.
```
Comment (jieqiboh) [inline docs/DeveloperGuide.md:566]: User inputs the command, specifying internship details

File: docs/DeveloperGuide.md [hunk 20]
```diff
@@ -885,6 +977,44 @@ Make sure to use the `exit` command or the close button to save data while closi
    
    2. To simulate a corrupted data file, edit the data file to contain some random text. Launch the app. The app should detect the corrupted file and automatically replace it with a new empty data file. You can then add new data to the app or reset the data to sample data by deleting the data file.
 
+--------------------------------------------------------------------------------------------------------------------
+
+## **Appendix: Design Decisions**
```
Comment (jieqiboh) [inline docs/DeveloperGuide.md:982]: Update header to be ###

File: docs/DeveloperGuide.md [hunk 21]
```diff
@@ -885,6 +977,44 @@ Make sure to use the `exit` command or the close button to save data while closi
...
+
+### Prefix-based Command Format
```
Comment (jieqiboh) [inline docs/DeveloperGuide.md:984]: Update header

**Review by jieqiboh — APPROVED** [review]

**Pull request merged**